java CDK: ContainerFactory properly supports unshared testcontainers

postamar opened the pull request:
Originally, the db-source's ContainerFactory test fixture was built under the assumption that testcontainers could always be shared by multiple tests. In practice it turns out that that's not always the case, especially for CDC tests where the CDC state is at the RDBMS level. For instance, in mysql, there's only one binlog.

Recently we struggled to debug an issue because an unshared testcontainer didn't forward its logs to the logging back-end. This is done by default for shared containers, but was forgotten for an unshared container in a CDC test. 

Rather than peppering these tests with the suitable logging incantations, this PR adds an `unshared` method to `ContainerFactory` which is like the existing `shared` except that, well, the container is not shared. This way testcontainers are provisioned in the same way regardless of whether they're shared or not.

This PR can be reviewed commit-by-commit.

stephane-airbyte commented on airbyte-cdk/java/airbyte-cdk/airbyte-commons/src/main/java/io/airbyte/commons/logging/LoggingHelper.java:
```diff
@@ -22,6 +22,7 @@ public enum Color {
     YELLOW_BACKGROUND("\u001b[43m"), // destination
```
<!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: line 13]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->I had no idea this existed. <!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34865?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

stephane-airbyte commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:
```diff
@@ -39,80 +29,14 @@ public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
-      }
-    }
-
+  default C unshared(String imageName, String... methods) {
```
exclusive?

postamar replied:
Let's go with `exclusive`.

stephane-airbyte commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:
```diff
@@ -0,0 +1,112 @@
+/**
+ * This class wraps a specific shared testcontainer instance, which is created exactly once.
+ */
+class ContainerFactoryWrapper<C extends JdbcDatabaseContainer<?>> {
```
does it need to be a JdbcDatabaseContainer? Or can we use GenericContainer here?

postamar replied:
Indeed we can.

stephane-airbyte commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:
```diff
@@ -0,0 +1,112 @@
+      .setPrefixColor(LoggingHelper.Color.RED_BACKGROUND);
+
+  @SuppressWarnings("unchecked")
+  static <C extends JdbcDatabaseContainer<?>> C getOrCreateShared(ContainerFactory<C> factory, String imageName, String... methods) {
```
That's super weird to have a different C here than in the class definition.

postamar replied:
It's static

stephane-airbyte replied:
I feel like all the methods here should be static, and the main class shouldn't be parameterized, no?

postamar replied:
You're right, it's better when the class isn't parameterized.

postamar commented:
@stephane-airbyte I addressed your comments. I noticed the CI still fails so I'll wait for https://github.com/airbytehq/airbyte/pull/34182 to merge before anything else.

postamar commented:
closed in favour of https://github.com/airbytehq/airbyte/pull/34892
